### PR TITLE
InConditionBuilder proper null handling

### PR DESCRIPTION
**Pull request by hstrychalski**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #17661

### CODE REVIEW

**Comment by samdark** [comment]
Seems to fail tests: https://travis-ci.org/yiisoft/yii2/jobs/618641372

**Comment by strychu** [comment]
I've fixed tests and added some test cases to condition provider

**Comment by strychu** [comment]
Although i see that build fails for php 7.4

**Comment by samdark** [comment]
That's OK. Same in master.

File: tests/framework/db/QueryBuilderTest.php [hunk 1]
```diff
@@ -1178,6 +1178,18 @@ public function conditionProvider()
 
             [['in', 'id', new TraversableObject([1, 2, 3])], '[[id]] IN (:qp0, :qp1, :qp2)', [':qp0' => 1, ':qp1' => 2, ':qp2' => 3]],
 
+            //in using array objects containing null value
+            [['in', 'id', new TraversableObject([1, null])], '[[id]]=:qp0 AND [[id]] IS NULL', [':qp0' => 1]],
```
Comment (samdark) [inline tests/framework/db/QueryBuilderTest.php]: That doesn't look right. `in` is about the value is being in the list. "in X [1,2,3]" transforms into `X IN [1,2,3]` that is equal to `X = 1 OR X = 2 OR X = 3`. Here we have `AND` and I think it should be `OR`.

File: tests/framework/db/QueryBuilderTest.php [hunk 2]
```diff
@@ -1178,6 +1178,18 @@ public function conditionProvider()
...
+            [['in', 'id', new TraversableObject([1, 2, null])], '[[id]] IN (:qp0, :qp1) AND [[id]] IS NULL', [':qp0' => 1, ':qp1' => 2]],
```
Comment (samdark) [inline tests/framework/db/QueryBuilderTest.php]: Same here. Looks like it should generate `OR`.

File: tests/framework/db/QueryBuilderTest.php [hunk 3]
```diff
@@ -1178,6 +1178,18 @@ public function conditionProvider()
...
+
+            //not in using array object containing null value
+            [['not in', 'id', new TraversableObject([1, null])], '[[id]]<>:qp0 AND [[id]] IS NOT NULL', [':qp0' => 1]],
```
Comment (samdark) [inline tests/framework/db/QueryBuilderTest.php]: These ones are correct.

File: tests/framework/db/QueryBuilderTest.php [hunk 4]
```diff
@@ -1178,6 +1178,18 @@ public function conditionProvider()
...
+            [['not in', 'id', new TraversableObject([1, 2, null])], '[[id]] NOT IN (:qp0, :qp1) AND [[id]] IS NOT NULL', [':qp0' => 1, ':qp1' => 2]],
+
+            //in using array object containing only null value
+            [['in', 'id', new TraversableObject([null])], '[[id]] IS NULL', []],
```
Comment (samdark) [inline tests/framework/db/QueryBuilderTest.php]: These ones are correct as well.

**Review by samdark — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**